src/base/threading.h

```diff
@@ -33,92 +33,76 @@
 
 #include <config.h>
 #include "base/basictypes.h"
+#include <thread>
 
-// Also allow for printing of a PerftoolsThreadID.
-#define GPRIuPTHREAD "lu"
-#define GPRIxPTHREAD "lx"
+// Also allow for printing of a std::thread::id.
+#define GPRIuTHREADID "zu"
+#define GPRIxTHREADID "zx"
 
-#if defined(__CYGWIN__) || defined(__CYGWIN32__) || defined(__APPLE__) || defined(__FreeBSD__)
-#  define PRINTABLE_PTHREAD(pthreadt) reinterpret_cast<uintptr_t>(pthreadt)
-#elif defined(__QNXNTO__)
-#  define PRINTABLE_PTHREAD(pthreadt) static_cast<intptr_t>(pthreadt)
-#else
-#  define PRINTABLE_PTHREAD(pthreadt) pthreadt
-#endif
+// Unfortunately std::thread::id does not provide a non-allocating,
+// non-throwing and generally "safe" means to print its value.
+// Nor does it provide a native_handle() function to get the underlying
+// platform-specific tid.
+// Doing the next best thing which is to print its hash value instead.
+#define PRINTABLE_THREADID(tid) std::hash<std::thread::id>()(tid)
 
 #ifdef _WIN32 // Should cover both toolchains on Windows - MSVC & MINGW
 
-using PerftoolsThreadID = DWORD;
-using PerftoolsTlsKey = DWORD;
-
-inline PerftoolsThreadID PerftoolsGetThreadID() {
-  return GetCurrentThreadId();
-}
-inline int PerftoolsThreadIDEquals(PerftoolsThreadID left, PerftoolsThreadID right) {
-  return left == right;
-}
-
-extern "C" PerftoolsTlsKey PthreadKeyCreate(void (*destr_fn)(void*));  /* port.cc */
-
-inline int PerftoolsCreateTlsKey(PerftoolsTlsKey *pkey, void (*destructor)(void*)) {
-  PerftoolsTlsKey key = PthreadKeyCreate(destructor);
-
-  if (key != TLS_OUT_OF_INDEXES) {
-    *(pkey) = key;
-    return 0;
-  }
-  else {
-    return GetLastError();
-  }
-}
-inline void* PerftoolsGetTlsValue(PerftoolsTlsKey key) {
-  DWORD err = GetLastError();
-  void* rv = TlsGetValue(key);
-
-  if (err) SetLastError(err);
-  return rv;
-}
-inline int PerftoolsSetTlsValue(PerftoolsTlsKey key, const void* value) {
-  if (TlsSetValue(key, (LPVOID)value)) {
-    return 0;
-  }
-  else {
-    return GetLastError();
-  }
-}
-
-inline void PerftoolsYield() {
-  Sleep(0);
-}
+namespace tcmalloc {
+
+    using TlsKey = DWORD;
+
+    extern "C" TlsKey PthreadKeyCreate(void (*destr_fn)(void*));  /* windows/port.cc */
```

Comment:
would be most sensible to remove extern "C" here.

Reply:
Done!